zk_prover/src/chips/merkle_sum_tree.rs

```diff
@@ -59,6 +59,22 @@ impl<const N_ASSETS: usize> MerkleSumTreeChip<N_ASSETS> {
             let element_l_next = meta.query_advice(col_a, Rotation::next());
             let element_r_next = meta.query_advice(col_b, Rotation::next());
 
+            // Audit: The constraint is aimed at checking the correct swap of the values. If swap_bit is 1,
+            // element_l_cur == element_r_next and element_r_cur == element_l_next.
+            // If swap_bit is 0, element_l_cur == element_l_next and element_r_cur == element_r_next.
+            // However, if we combine the two equations, there's a potential for unintended solutions
+            // that satisfy the composite equation without satisfying the intended individual constraints.
+            // For the case where swap_bit = 0:
+            // element_l_cur − element_l_next + element_r_next − element_r_cur = 0
+            // Here, it's theoretically possible to cheat by picking element_l_cur, element_l_next,
+            // element_r_cur, and element_r_next such that they don't individually satisfy
+            // element_l_cur − element_l_next = 0 and element_l_cur − element_l_next = 0
+            // but still satisfy the combined equation.
+            // For example, if element_l_cur = 3 and element_l_next = 1, and element_r_cur = 2 and element_r_next = 4,
+            // the composite equation would still be satisfied because:
+            // 3 − 1 + 4 − 2 = 0
```

Comment:
I understand the issue you've pointed out. A malicious prover could potentially exploit the switch in values at this point:
https://github.com/summa-dev/summa-solvency/blob/7489c53ebdd4dd8bbcb3b45490cccee6426fc7c5/zk_prover/src/chips/merkle_sum_tree.rs#L223-L235

However, considering the example provided, it seems that the constraint would not be satisfied.

$$ 1 * 2 * ( 3 - 2 ) - ( 1 - 3) - ( 2 - 4) = 2 \neq 0 $$

Please correct me if I'm missing something or if I'm wrong.

Reply:
The example is for the case when the swap bit is zero: swap_bit * 2 * (element_r_cur - element_l_cur) - (element_l_next - element_l_cur) - (element_r_cur - element_r_next) = - (element_l_next - element_l_cur) - (element_r_cur - element_r_next) = element_l_cur - element_l_next + element_r_next - element_r_cur

Reply:
For example, 3 - 4 + 2 - 1 = 0.

Reply:
I see. For better understanding in the example, perhaps assigning 5 to `element_l_next` instead of 1 would make it clearer.

Reply:
Yes, I think this is the solution.

Reply:
Agreed